File: java/P22.java
// P22 Create a list containing all integers within a given range.

import core.List;

import static core.List.cons;
import static core.List.nil;

class P22 {
    // Recursion
    List<Integer> f1(int a, int b) {
        return a <= b ? cons(a, f1(a + 1, b)) : List.<Integer>nil();
    }

    // Tail recursion
    List<Integer> f2(int a, int b) {
        return f2_inner(a, b, List.<Integer>nil());
    }

    private List<Integer> f2_inner(int a, int b, List<Integer> acc) {
        return b >= a ? f2_inner(a, b - 1, cons(b, acc)) : acc;
    }

    // Imperative
    List<Integer> f3(int a, int b) {
        int i = b;
        List<Integer> result = nil();
        while (i >= a) {
            result = cons(i, result);
            i--;
        }
        return result;
    }
}

File: scala/P22.scala
// P22 Create a list containing all integers within a given range.

// Through predefined "Range"
def f1(a: Int, b: Int): List[Int] =
  (a to b).toList

// A part of an infinite list
def f2(a: Int, b: Int): List[Int] =
  Stream.iterate(a)(1 +).take(b - a + 1).toList

// As a sum
def f3(a: Int, b: Int): List[Int] =
  List.fill(b - a)(1).scan(a)(_ + _)

// Recursion
def f4(a: Int, b: Int): List[Int] =
  if (a <= b) a :: f4(a + 1, b) else Nil

// Tail recursion
@annotation.tailrec
def f5(a: Int, b: Int, acc: List[Int] = Nil): List[Int] =
  if (b >= a) f5(a, b - 1, b :: acc) else acc

// Imperative
def f6(a: Int, b: Int): List[Int] = {
  var i = b
  var result: List[Int] = Nil
  while (i >= a) {
    result ::= i
    i -= 1
  }
  result
}